File: tests/docs/python/test_attribute_read.py
# Copyright (c) 2025-2026, NVIDIA CORPORATION. All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto.  Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.

"""Tests for ovstage attribute reads used with an attached renderer.

Reads target schema-known attributes the runtime exposes to Fabric:
``omni:rtx:rtpt:maxBounces`` on the RenderProduct (a 32-bit integer; the
runtime may report it as either int32 or uint32) and ``points`` on
``/World/Plane`` (a float3 array authored in ``ovrtx-test-base-geometry.usda``).
"""

from pathlib import Path

import numpy as np
import ovrtx
import ovstage
import pytest

TEST_BASE_PATH = str((Path(__file__).parent / "../data/ovrtx-test-base.usda").resolve())


def _load_base(stage):
    ovstage.population.open_usd(stage, TEST_BASE_PATH, ordinal=1)
    stage.advance_write_floor(1, ovstage.Scope.ALL).wait()


def _load_base_legacy(renderer):
    renderer.open_usd(TEST_BASE_PATH)
    renderer.reset()


def test_read_scalar_attribute_cpu(stage):
    """Write a schema-known int32 scalar and read it back."""
    _load_base(stage)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/Render/Camera"])
        query = stage.query_from_path_list(path_list)
        attribute = paths.intern_token("omni:rtx:rtpt:maxBounces")
        stage.write_attribute(
            query, attribute, ordinal=2, tensors=np.array([17], dtype=np.uint32), is_array=False
        ).wait()
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()

        # [snippet:doc-read-attribute-scalar]
        read = stage.read_attributes(query, [attribute], ovstage.OrdinalRange.latest(2))
        read.wait()
        group = read.fetch_next()
        tensor = group.dlpack(0)
        values = np.from_dlpack(tensor).copy()
        stage.release_group(group)
        read.release().wait()
        # [/snippet:doc-read-attribute-scalar]

        query.release().wait()
        paths.destroy_path_list(path_list)

    assert isinstance(tensor, ovstage.ManagedDLTensor)
    # maxBounces is a 32-bit integer; the runtime may report it as int32 or
    # uint32 depending on the schema — compare the value, not the sign-ness.
    assert values.dtype.itemsize == 4 and values.dtype.kind in ("i", "u")
    assert values.shape == (1,)
    assert int(values[0]) == 17


@pytest.mark.filterwarnings("ignore:.* is deprecated in ovrtx 0\\.4\\..*:DeprecationWarning")
def test_read_scalar_attribute_into_dest(renderer):
    """Pass a pre-allocated tensor as ``dest``; verify data lands in it."""
    _load_base_legacy(renderer)

    renderer.write_attribute(
        prim_paths=["/Render/Camera"],
        attribute_name="omni:rtx:rtpt:maxBounces",
        tensor=np.array([11], dtype=np.int32),
    )

    # [snippet:doc-read-attribute-dest-tensor]
    # Pre-allocate the destination. The read writes directly into `dest`; the
    # returned tensor is a handle to the same memory — both aliases are valid.
    # The dtype must match how the runtime stores the attribute.
    dest = np.empty((1,), dtype=np.uint32)
    renderer.read_attribute(
        attribute_name="omni:rtx:rtpt:maxBounces",
        prim_paths=["/Render/Camera"],
        dest=dest,
    )
    # `dest` now holds the attribute value.
    # [/snippet:doc-read-attribute-dest-tensor]

    assert int(dest[0]) == 11


def test_read_array_attribute(stage):
    """Read ``points`` (``float3[]``) from the Plane mesh."""
    _load_base(stage)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/World/Plane"])
        query = stage.query_from_path_list(path_list)
        points = paths.intern_token("points")

        # [snippet:doc-read-array-attribute]
        with stage.read_attributes(query, [points], ovstage.OrdinalRange.latest(1)) as read:
            group = read.fetch_next()
            values = np.from_dlpack(group.dlpack(0)).copy()
            stage.release_group(group)
        # [/snippet:doc-read-array-attribute]

        query.release().wait()
        paths.destroy_path_list(path_list)

    # ovrtx-test-base-geometry.usda authors 4 float3 points on the Plane.
    assert values.dtype == np.float32
    assert values.size == 4 * 3


def test_read_attribute_async(stage):
    """Exercise the two-phase async lifecycle for an attribute read."""
    _load_base(stage)

    with ovstage.PathDictionary(stage) as paths:
        path_list = paths.create_path_list_from_strings(["/Render/Camera"])
        query = stage.query_from_path_list(path_list)
        attribute = paths.intern_token("omni:rtx:rtpt:maxBounces")
        stage.write_attribute(
            query, attribute, ordinal=2, tensors=np.array([5], dtype=np.uint32), is_array=False
        ).wait()
        stage.advance_write_floor(2, ovstage.Scope.ALL).wait()

        # [snippet:doc-read-attribute-async]
        read = stage.read_attributes(query, [attribute], ovstage.OrdinalRange.latest(2))
        read.wait()
        group = read.fetch_next()
        values = np.from_dlpack(group.dlpack(0)).copy()
        stage.release_group(group)
        read.release().wait()
        # [/snippet:doc-read-attribute-async]

        query.release().wait()
        paths.destroy_path_list(path_list)

    assert int(values[0]) == 5


@pytest.mark.filterwarnings("ignore:.* is deprecated in ovrtx 0\\.4\\..*:DeprecationWarning")
def test_read_attribute_cuda_dest(renderer):
    """Read directly into a GPU (CUDA) destination tensor via DLPack."""
    import warp as wp  # via warp-lang; provides DLPack-compatible CUDA arrays

    _load_base_legacy(renderer)

    renderer.write_attribute(
        prim_paths=["/Render/Camera"],
        attribute_name="omni:rtx:rtpt:maxBounces",
        tensor=np.array([9], dtype=np.int32),
    )

    # [snippet:doc-read-attribute-cuda-dest]
    # Allocate a CUDA destination through Warp (any DLPack-compatible CUDA
    # allocator works). The read writes directly into GPU memory; pass a CUDA
    # stream handle so the read is ordered on the caller's stream.
    dest = wp.empty(1, dtype=wp.uint32, device="cuda:0")
    stream = wp.Stream(device=dest.device)
    renderer.read_attribute(
        attribute_name="omni:rtx:rtpt:maxBounces",
        prim_paths=["/Render/Camera"],
        dest=dest,
        cuda_stream=stream.cuda_stream,
    )
    wp.synchronize_stream(stream)
    # [/snippet:doc-read-attribute-cuda-dest]

    # Copy back to host and verify the value.
    host = dest.numpy()
    assert int(host[0]) == 9

File: tests/docs/c/test_attribute_read.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
// property and proprietary rights in and to this material, related
// documentation and any modifications thereto. Any use, reproduction,
// disclosure or distribution of this material and related documentation
// without an express license agreement from NVIDIA CORPORATION or
// its affiliates is strictly prohibited.

// Tests for ovstage attribute reads used with an attached renderer. Mirrors
// tests/docs/python/test_attribute_read.py.
// - Scalar test: write-then-read omni:rtx:rtpt:maxBounces on /Render/Camera.
// - Array test: read the authored "points" attribute from /World/Plane.

#include <gtest/gtest.h>
#include "helpers.h"

#include <cstdint>
#include <cstring>
#include <string>
#include <vector>

class AttributeReadTest : public DocsOvstageTestBase {
protected:
    DOCS_OVSTAGE_TEST_SUITE(AttributeReadTest)
};

TEST_F(AttributeReadTest, ReadScalarAttribute) {
    docs_load_base();
    if (HasFatalFailure()) return;

    // Setup a single-prim query on /Render/Camera and intern the attribute
    // token. Both are kept alive across the write and the read.
    path_dictionary_instance_t* pd = ovstage_get_path_dictionary(stage_);
    ASSERT_NE(pd, nullptr);

    ovx_string_t prim_path = ovx_str("/Render/Camera");
    ovx_primpath_list_t path_list{};
    ASSERT_EQ(path_dictionary_create_path_list_from_strings(pd, &prim_path, 1, &path_list).status,
              OVX_API_SUCCESS);

    ovstage_query_handle_t query_handle = OVSTAGE_INVALID_QUERY_HANDLE;
    ASSERT_EQ(ovstage_query_from_path_list(stage_, path_list, &query_handle), OVSTAGE_OK)
        << format_ovstage_last_error();

    ovx_string_t attr_name = ovx_str("omni:rtx:rtpt:maxBounces");
    ovx_token_t attr_token{};
    ASSERT_EQ(path_dictionary_create_tokens_from_strings(pd, &attr_name, 1, &attr_token).status,
              OVX_API_SUCCESS);

    // Write 17 to maxBounces at ordinal 2, then seal so a subsequent read at
    // latest(2) has committed data to observe.
    uint32_t write_value = 17;
    int64_t write_shape[1] = {1};
    DLTensor write_tensor{};
    write_tensor.data = &write_value;
    write_tensor.device = {kDLCPU, 0};
    write_tensor.ndim = 1;
    write_tensor.dtype = {kDLUInt, 32, 1};
    write_tensor.shape = write_shape;
    write_tensor.strides = nullptr;
    write_tensor.byte_offset = 0;

    ovstage_write_data_t write_data{};
    write_data.tensors = &write_tensor;
    write_data.tensor_count = 1;
    write_data.is_array = false;

    ovx_string_or_token_t attr_ref{};
    attr_ref.token = attr_token;

    ovstage_enqueue_result_t wq = ovstage_write_attribute(
        stage_, query_handle, attr_ref, /*ordinal=*/2, write_data, OVSTAGE_PRIM_MODE_UPSERT);
    ASSERT_EQ(wq.status, OVSTAGE_OK) << format_ovstage_last_error();
    docs_wait_ovstage_no_errors(stage_, wq.op_index);
    docs_ovstage_advance_write_floor(stage_, 2);

    // [snippet:doc-read-attribute-scalar-c]
    // Enqueue a read for a schema-known attribute over one prim, at latest(2).
    // The read handle is reserved synchronously — pass it to fetch_read_next
    // once the returned op_index completes.
    ovstage_ordinal_range_t range{};
    range.has_start_ordinal = false;
    range.end_ordinal = 2;

    ovstage_read_handle_t read_handle = OVSTAGE_INVALID_READ_HANDLE;
    ovstage_enqueue_result_t eq = ovstage_read_attributes(
        stage_, query_handle, &attr_token, 1, range, &read_handle);
    ASSERT_EQ(eq.status, OVSTAGE_OK) << format_ovstage_last_error();
    docs_wait_ovstage_no_errors(stage_, eq.op_index);

    // Fetch the first (and, for a single-prim scalar read, only) group. The
    // group's tensor is a DLPack view into the sealed attribute storage —
    // valid until release_group.
    ovstage_read_group_t group{};
    ASSERT_EQ(ovstage_fetch_read_next(stage_, read_handle, OVSTAGE_TIMEOUT_INFINITE, &group),
              OVSTAGE_OK)
        << format_ovstage_last_error();
    ASSERT_GT(group.data.tensor_count, 0u);
    uint32_t value = *static_cast<uint32_t const*>(group.data.tensors[0].data);

    // Release the fetched storage (synchronous) then the read handle (async;
    // the release is per-handle-ordered so any in-flight fetch drains first).
    ovstage_release_group(stage_, &group);
    ovstage_release_read(stage_, read_handle);
    // [/snippet:doc-read-attribute-scalar-c]

    EXPECT_EQ(value, 17u);

    ovstage_release_query(stage_, query_handle);
    path_dictionary_release_path_list_reference(pd, path_list);
}

TEST_F(AttributeReadTest, ReadArrayAttribute) {
    docs_load_base();
    if (HasFatalFailure()) return;

    // Setup: single-prim query on /World/Plane and the "points" token.
    path_dictionary_instance_t* pd = ovstage_get_path_dictionary(stage_);
    ASSERT_NE(pd, nullptr);

    ovx_string_t prim_path = ovx_str("/World/Plane");
    ovx_primpath_list_t path_list{};
    ASSERT_EQ(path_dictionary_create_path_list_from_strings(pd, &prim_path, 1, &path_list).status,
              OVX_API_SUCCESS);

    ovstage_query_handle_t query_handle = OVSTAGE_INVALID_QUERY_HANDLE;
    ASSERT_EQ(ovstage_query_from_path_list(stage_, path_list, &query_handle), OVSTAGE_OK)
        << format_ovstage_last_error();

    ovx_string_t attr_name = ovx_str("points");
    ovx_token_t attr_token{};
    ASSERT_EQ(path_dictionary_create_tokens_from_strings(pd, &attr_name, 1, &attr_token).status,
              OVX_API_SUCCESS);

    // [snippet:doc-read-array-attribute-c]
    // Array attributes are variable-length per prim. Read semantics are the
    // same as for scalars — enqueue, wait, fetch — but the DLTensor's shape
    // reflects the per-prim element count; for `points` (float3 array) the
    // dtype.lanes carries the tuple width and the leading shape dim carries
    // the element (point) count.
    ovstage_ordinal_range_t range{};
    range.has_start_ordinal = false;
    range.end_ordinal = 1;

    ovstage_read_handle_t read_handle = OVSTAGE_INVALID_READ_HANDLE;
    ovstage_enqueue_result_t eq = ovstage_read_attributes(
        stage_, query_handle, &attr_token, 1, range, &read_handle);
    ASSERT_EQ(eq.status, OVSTAGE_OK) << format_ovstage_last_error();
    docs_wait_ovstage_no_errors(stage_, eq.op_index);

    ovstage_read_group_t group{};
    ASSERT_EQ(ovstage_fetch_read_next(stage_, read_handle, OVSTAGE_TIMEOUT_INFINITE, &group),
              OVSTAGE_OK)
        << format_ovstage_last_error();
    ASSERT_GT(group.data.tensor_count, 0u);
    ASSERT_TRUE(group.is_array);

    DLTensor const& t = group.data.tensors[0];
    // ovrtx-test-base-geometry.usda authors 4 float3 points on the Plane.
    ASSERT_EQ(t.dtype.code, kDLFloat);
    ASSERT_EQ(t.dtype.bits, 32u);
    ASSERT_EQ(t.dtype.lanes, 3u);
    ASSERT_EQ(t.ndim, 1);
    ASSERT_EQ(t.shape[0], 4);
    int64_t element_count = t.shape[0] * t.dtype.lanes;

    ovstage_release_group(stage_, &group);
    ovstage_release_read(stage_, read_handle);
    // [/snippet:doc-read-array-attribute-c]

    EXPECT_EQ(element_count, 12);

    ovstage_release_query(stage_, query_handle);
    path_dictionary_release_path_list_reference(pd, path_list);
}
